python/ray/autoscaler/launch_and_verify_cluster.py

```diff
@@ -229,7 +231,49 @@ def cleanup_cluster(cluster_config):
     raise last_error
 
 
-def run_ray_commands(cluster_config, retries, no_config_cache, num_expected_nodes=1):
+def cleanup_security_group(ec2_client, id):
+    retry = 0
+    while retry < 10:
+        try:
+            ec2_client.delete_security_group(GroupId=id)
+            return
+        except botocore.exceptions.ClientError as e:
+            if e.response["Error"]["Code"] == "DependencyViolation":
+                sleep_time = 2**retry
+                print(
+                    f"Waiting {sleep_time}s for the instance to be terminated before deleting the security group {id}"  # noqa E501
+                )
+                time.sleep(sleep_time)
+                retry += 1
+            else:
+                print(f"Error deleting security group: {e}")
+                return
```

Comment:
It looks like we don't raise an exception if deleting the security groups fails.  That's probably fine, since it's not essential for the test and we don't want the release test to become too flaky.

On the other hand, if deleting security groups fails 100% of the time, how would we ever detect it? I guess we could manually check the logs, or in the worst case, we would just accumulate. security groups and hit the limit again in a few months, which maybe isn't so bad.  I don't think it needs to block the PR, but it would be good to get your thoughts.

Reply:
Yeah. You are right. We don't raise an exception here because we can't guarantee 100% that security groups are deleted. There might be network issues, AWS issues, etc. We don't want these factors to make the release test flaky.

The bottom line is that we can find the root cause if something goes wrong -- for example, security groups were dangling again. In such case, some logs are good enough.